aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/ECSComputeService.java

```diff
@@ -84,22 +92,63 @@ protected ECSComputeService(ComputeServiceContext context, Map<String, Credentia
                                Optional<SecurityGroupExtension> securityGroupExtension,
                                DelegatingImageExtension.Factory delegatingImageExtension) {
       super(context, credentialStore, images, sizes, locations, listNodesStrategy, getImageStrategy,
-            getNodeMetadataStrategy, runNodesAndAddToSetStrategy, rebootNodeStrategy, destroyNodeStrategy,
-            startNodeStrategy, stopNodeStrategy, templateBuilderProvider, templateOptionsProvider, nodeRunning,
-            nodeTerminated, nodeSuspended, initScriptRunnerFactory, initAdminAccess, runScriptOnNodeFactory,
-            persistNodeCredentials, userExecutor, imageExtension, securityGroupExtension, delegatingImageExtension);
+              getNodeMetadataStrategy, runNodesAndAddToSetStrategy, rebootNodeStrategy, destroyNodeStrategy,
+              startNodeStrategy, stopNodeStrategy, templateBuilderProvider, templateOptionsProvider, nodeRunning,
+              nodeTerminated, nodeSuspended, initScriptRunnerFactory, initAdminAccess, runScriptOnNodeFactory,
+              persistNodeCredentials, userExecutor, imageExtension, securityGroupExtension, delegatingImageExtension);
       this.cleanupResources = cleanupResources;
    }
 
    @Override
    protected void cleanUpIncidentalResourcesOfDeadNodes(Set<? extends NodeMetadata> deadNodes) {
       for (NodeMetadata deadNode : deadNodes) {
+         RegionAndId regionAndId = RegionAndId.fromSlashEncoded(deadNode.getId());
+         Set<String> tags = deadNode.getTags();
+         String vSwitchId = extractVSwitchId(tags);
+         VSwitch vSwitch = context.unwrapApi(ECSComputeServiceApi.class).vSwitchApi().list(deadNode.getLocation().getId(), ListVSwitchesOptions.Builder.vSwitchId(vSwitchId)).first().orNull();
+         String vpcId = vSwitch.vpcId();
+
          try {
-            cleanupResources.cleanupNode(deadNode.getId());
+            cleanupResources.cleanupNode(regionAndId);
          } catch (Exception ex) {
             logger.warn(ex, "Error cleaning up resources for node %s", deadNode);
          }
+
+         List<SecurityGroup> securityGroups = cleanupResources.findOrphanedSecurityGroups(regionAndId.regionId(), deadNode.getGroup());
+         for (SecurityGroup securityGroup : securityGroups) {
+            logger.debug(">> destroying security group %s ...", securityGroup.id());
+            if (cleanupResources.cleanupSecurityGroupIfOrphaned(regionAndId.regionId(), securityGroup.id())) {
+               logger.debug(">> security group: (%s) has been deleted.", securityGroup.id());
+            } else {
+               logger.warn(">> security group: (%s) has not been deleted.", securityGroup.id());
+            }
+         }
+
+         // FIXME not sure it is correct to always delete vSwitch and VPC
```

Comment:
@nacx wdyt? I'm not completely sure about this default behaviour

Reply:
This is a similar case as this (still unfinished) PR in ARM provider. 

Basically we should be able to identify if those VPC/vSwitch were created by the jclouds providers and delete them only if empty AND created by us. We use tags in the ARM to check this, is there anything similar in Aliyun API?

Reply:
Good point @danielestevez unfortunately tagApi don't work for VPC or VSwitch, only instances, images and few others. Trying to use the `description` although super ugly

Reply:
thinking about it more, regardless of the tags, I think the question is whether is right to destroy a VPC every time jclouds deletes a group of nodes or not. I honestly don't know the answer as I can see value in both cases, so maybe an extra property configurable by the user?

Reply:
I don't see the value in having jclouds deleting VPCs that were not created by jclouds. We should always keep the resources created by the user by the provider API/Portal or 3rd party apps
What case of use would make sense here?

Reply:
+1

I was thinking about giving the opportunity to decide whether to keep a VPC created by jclouds or delete it for every deployment.

Reply:
I'd say we want to delete the VPCs *we created* if they are no longer in use. When the last node is removed, then delete the VPC and vSwitch, if we created them. To do that, though, we need to be able to identify them, and the description and name fields seem flaky.

Some questions:
* Are users charged by VPCs, vSwitches?
* Is there a limit on the number of VPCs?
* Other concerns that may have a direct impact on users?

Reply:
flaky, indeed!

From the doc:
- costs: I think it's usage is for free
- Maximum number of VPCs per region: 10 (Submit a ticket to apply for more quota)
- other concerns: although fast, creating/destroying VPC adds a delay, but not crucial